uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,167 @@
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def _is_any_beta(cfg: UAConfig, ents: List[str]) -> Tuple[bool, str]:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True, name
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False, ""
+
+
+def _is_compatible_services_present(
+    cfg, ent_list: List[str]
+) -> Tuple[bool, str, List[str]]:
+    ent_cls = list()  # type: List[Type[UAEntitlement]]
+    for e in ent_list:
+        ent_cls.append(entitlements.entitlement_factory(cfg, e))
+
+    for ent in ent_cls:
+        if ent().incompatible_services:
+            incompat = [
+                service.entitlement for service in ent().incompatible_services
+            ]
+            if any(e in incompat for e in ent_cls):
+                return True, ent().name, [e().name for e in incompat]
+    return False, "", []
+
+
+def full_auto_attach(options: FullAutoAttachOptions) -> FullAutoAttachResult:
+    return _full_auto_attach(options, UAConfig(root_mode=True))
+
+
+def _full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    event.set_event_mode(event_logger.EventLoggerMode.JSON)
+
+    services = set()
+    if options.enable:
+        is_beta_found, service = _is_any_beta(cfg, options.enable)
+        if is_beta_found:
+            raise exceptions.BetaServiceError(
+                msg="beta service found in the enable list",
+                msg_code="beta-service-found",
+                additional_info={"beta_service": service},
+            )
+        services.update(options.enable)
+    if options.enable_beta:
+        services.update(options.enable_beta)
+
+    service_list = list(services)
+
+    found, not_found = entitlements.get_valid_entitlement_names(
+        service_list, cfg
+    )
+
+    incompat_detected, ent, incompat_ents = _is_compatible_services_present(
+        cfg, sorted(found)
+    )  # sort for easy testing exceptions raised
+    if incompat_detected:
+        msg_ = "{ent} is incompatible with any of these services {incompat}"
+        raise exceptions.IncompatibleEntitlementsDetected(
+            msg=msg_.format(ent=ent, incompat=incompat_ents),
+            msg_code="incompatible-services-detected",
+            additional_info={
+                "service": ent,
+                "incompatible_services": ",".join(incompat_ents),
+            },
+        )
+
+    instance = actions.get_cloud_instance(cfg)
+    enable_default_services = (
+        options.enable is None and options.enable_beta is None
+    )
+    token = _get_auto_attach_token(cfg, instance)  # type: ignore
+    actions.attach_with_token(
+        cfg, token=token, allow_enable=enable_default_services
+    )
```

Comment:
Can this be replaced with
```suggestion
    actions.auto_attach(
        cfg, instance, allow_enable=enable_default_services
    )
```

Reply:
Oh, yeah. I am trying to remember why I created `_get_auto_attach_token`, it seems I was trying to solve an issue that currently is not there.

Reply:
Oh I have just remembered, it is because in `actions.auto_attach` the `allow_enable` parameter for `attach_with_token` function defaults to true. If I change it, i.e. add `allow_enable` parameter to `actions.auto_attach`, I will have to change more places. Could I create a PR for this change?

Reply:
@CalvoM 
I think the change is just this?
```diff
diff --git a/uaclient/actions.py b/uaclient/actions.py
index def703f1..aa70786e 100644
--- a/uaclient/actions.py
+++ b/uaclient/actions.py
@@ -71,7 +71,7 @@ def attach_with_token(
 
 
 def auto_attach(
-    cfg: config.UAConfig, cloud: clouds.AutoAttachCloudInstance
+    cfg: config.UAConfig,
+    cloud: clouds.AutoAttachCloudInstance,
+    allow_enable: bool = True
 ) -> None:
     """
     :raise UrlError: On unexpected connectivity issues to contract
@@ -95,7 +95,7 @@ def auto_attach(
 
     token = tokenResponse["contractToken"]
 
-    attach_with_token(cfg, token=token, allow_enable=True)
+    attach_with_token(cfg, token=token, allow_enable=allow_enable)
 
 
 def enable_entitlement_by_name(

```

and all tests should still pass with that change.


But if you'd like to do it in a follow up PR that's okay too :) (maybe I'm missing something)